pkg/operator/controller/garden/access/reconciler.go

```diff
@@ -0,0 +1,132 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package access
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kubernetesclient "github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/operator/apis/config"
+	operatorclient "github.com/gardener/gardener/pkg/operator/client"
+)
+
+// Reconciler reconciles garden access secrets.
+type Reconciler struct {
+	Client  client.Client
+	Manager manager.Manager
+	Config  *config.OperatorConfiguration
+	FS      afero.Fs
+
+	tokenFilePath  string
+	virtualCluster cluster.Cluster
+}
+
+// Reconcile processes the given access secret in the request.
+// It extracts the included Kubeconfig, and prepares a dedicated REST config
+// where the inline bearer token is replaced by a bearer token file.
+// Any subsequent reconciliation run causes the content of the BearerTokenFile to be updated with the token found
+// in the access secret.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, request.NamespacedName, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	kubeConfigRaw, ok := secret.Data[kubernetesclient.KubeConfig]
+	if !ok {
+		log.Info("Secret does not contain kubeconfig")
+		return reconcile.Result{}, nil
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfigRaw)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error creating REST config: %w", err)
+	}
+
+	if restConfig.BearerToken == "" {
+		log.Info("BearerToken is not yet populated in kubeconfig")
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.writeTokenToFile(restConfig.BearerToken); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error writing bearer token to file: %w", err)
+	}
+
+	return reconcile.Result{}, r.createVirtualCluster(log, restConfig)
+}
+
+// GetVirtualCluster returns the virtual cluster object.
+func (r *Reconciler) GetVirtualCluster() cluster.Cluster {
+	return r.virtualCluster
+}
+
+// CreateTemporaryFile creates a temporary file. Exposed for testing.
```

Comment:
I guess we discussed this some weeks back, but I forgot again. Can you please explain (and augment the comment) why we need this (and cannot just use the `MemOs` from `afero` for testing)?

Reply:
I adjusted the test to use `MemOs`.  `CreateTemporaryFile` is still in place to influence the file path which is otherwise unknown for temporary files.